### ISSUE TITLE
.envrc a symlink does not throw error

### ISSUE THREAD

**Issue by brikou** [direnv/direnv#288]
I want to easily switch .envrc (based on environment in my case).

Here is the way to reproduce my issue...

```sh
cd $(mktemp -d)

cat << 'EOF' > .envrc--production
export ENV=production
EOF

cat << 'EOF' > .envrc--development
export ENV=development
EOF

ln --symbolic --force .envrc--development .envrc
# direnv: error .envrc is blocked. Run `direnv allow` to approve its content.

direnv allow

echo $ENV
# development
```

If I want to switch env to production I then run:

```sh
ln --symbolic --force .envrc--production .envrc
# (...no error thrown)

echo $ENV
# development (instead of production, ... nothing loaded)
```

No error is throw and $ENV is still set to development. I still can force it with `direnv reload`, but it is ugly.

WDYT, is it a bug or a feature :)

**Comment by zimbatm** [comment]
Hey it's a cool idea. I would like to treat this as a bug :)

Can you run `direnv status` before and after the switch and also `stat .envrc--development .envrc--production`?

**Comment by brikou** [comment]
You can safely copy/paste the following snippet on your own, and so be able to reproduce the whole thing.

I can wait to have this ~~feature available~~, bug solved. BTW direnv is amazing!

```sh
cd $(mktemp -d)

cat << 'EOF' > .envrc--production
export ENV=production
EOF

cat << 'EOF' > .envrc--development
export ENV=development
EOF

ln --symbolic --force .envrc--development .envrc
# direnv: error .envrc is blocked. Run `direnv allow` to approve its content.

direnv allow

echo $ENV
# development

direnv status
# direnv exec path /usr/local/bin/direnv
# DIRENV_CONFIG /home/brikou/.config/direnv
# DIRENV_BASH /bin/bash
# Loaded RC path /tmp/tmp.5tvGz3Yoyq/.envrc
# Loaded watch: ".envrc" - 2017-08-12T15:28:21+02:00
# Loaded watch: "../../home/brikou/.config/direnv/allow/be0ef2f322890e4e7fd2d29c957c996f9d66517f837b46059bfe4a0c9cd88ed2" - 2017-08-12T15:28:21+02:00
# Loaded watch: "../../home/brikou/.direnvrc" - 2017-08-12T14:04:03+02:00
# Loaded RC allowed false
# Loaded RC allowPath 
# Found RC path /tmp/tmp.5tvGz3Yoyq/.envrc
# Found watch: ".envrc" - 2017-08-12T15:28:21+02:00
# Found watch: "../../home/brikou/.config/direnv/allow/be0ef2f322890e4e7fd2d29c957c996f9d66517f837b46059bfe4a0c9cd88ed2" - 2017-08-12T15:28:21+02:00
# Found RC allowed true
# Found RC allowPath /home/brikou/.config/direnv/allow/be0ef2f322890e4e7fd2d29c957c996f9d66517f837b46059bfe4a0c9cd88ed2

stat .envrc--development
# Fichier : '.envrc--development'
#  Taille : 23        	Blocs : 8          Blocs d'E/S : 4096   fichier
# Périphérique : 801h/2049d	Inœud : 16003621    Liens : 1
# Accès : (0664/-rw-rw-r--)  UID : ( 1000/  brikou)   GID : ( 1000/  brikou)
# Accès : 2017-08-12 15:28:21.235689736 +0200
# Modif. : 2017-08-12 15:28:21.227689982 +0200
# Changt : 2017-08-12 15:28:21.227689982 +0200
# Créé : -

stat .envrc--production
# Fichier : '.envrc--production'
#  Taille : 22        	Blocs : 8          Blocs d'E/S : 4096   fichier
# Périphérique : 801h/2049d	Inœud : 16003620    Liens : 1
# Accès : (0664/-rw-rw-r--)  UID : ( 1000/  brikou)   GID : ( 1000/  brikou)
# Accès : 2017-08-12 15:28:21.223690106 +0200
# Modif. : 2017-08-12 15:28:21.223690106 +0200
# Changt : 2017-08-12 15:28:21.223690106 +0200
# Créé : -

ln --symbolic --force .envrc--production .envrc
# (...no error thrown)

echo $ENV
# development (instead of production, ... nothing loaded)

direnv status
# direnv exec path /usr/local/bin/direnv
# DIRENV_CONFIG /home/brikou/.config/direnv
# DIRENV_BASH /bin/bash
# Loaded RC path /tmp/tmp.5tvGz3Yoyq/.envrc
# Loaded watch: ".envrc" - 2017-08-12T15:28:21+02:00
# Loaded watch: "../../home/brikou/.config/direnv/allow/be0ef2f322890e4e7fd2d29c957c996f9d66517f837b46059bfe4a0c9cd88ed2" - 2017-08-12T15:28:21+02:00
# Loaded watch: "../../home/brikou/.direnvrc" - 2017-08-12T14:04:03+02:00
# Loaded RC allowed false
# Loaded RC allowPath 
# Found RC path /tmp/tmp.5tvGz3Yoyq/.envrc
# Found watch: ".envrc" - 2017-08-12T15:28:21+02:00
# Found watch: "../../home/brikou/.config/direnv/allow/a59a8f9f93952cba7cc0eed246dae625d6c075de1d5a3657f6f5d1a6f54d9af8" - 1970-01-01T01:00:00+01:00
# Found RC allowed false
# Found RC allowPath /home/brikou/.config/direnv/allow/a59a8f9f93952cba7cc0eed246dae625d6c075de1d5a3657f6f5d1a6f54d9af8

stat .envrc--production
#   Fichier : '.envrc--production'
#    Taille : 22        	Blocs : 8          Blocs d'E/S : 4096   fichier
# Périphérique : 801h/2049d	Inœud : 16003620    Liens : 1
# Accès : (0664/-rw-rw-r--)  UID : ( 1000/  brikou)   GID : ( 1000/  brikou)
# Accès : 2017-08-12 15:31:17.665920690 +0200
# Modif. : 2017-08-12 15:28:21.223690106 +0200
# Changt : 2017-08-12 15:28:21.223690106 +0200
#   Créé : -
  
stat .envrc--development 
#   Fichier : '.envrc--development'
#    Taille : 23        	Blocs : 8          Blocs d'E/S : 4096   fichier
# Périphérique : 801h/2049d	Inœud : 16003621    Liens : 1
# Accès : (0664/-rw-rw-r--)  UID : ( 1000/  brikou)   GID : ( 1000/  brikou)
# Accès : 2017-08-12 15:28:21.235689736 +0200
# Modif. : 2017-08-12 15:28:21.227689982 +0200
# Changt : 2017-08-12 15:28:21.227689982 +0200
#   Créé : -

```

**Comment by brikou** [comment]
@zimbatm if .envrc (the symlink file) is remove before running ln -s, then it work as exepected. The problem comes with the `--force` arg which make something special, it is like direnv doesn't have the time to understand that the file is deleted

```sh
# this work
[ -e .envrc ] && rm .envrc
ln --symbolic .envrc--production .envrc

# but not this
ln --symbolic --force .envrc--production .envrc
```

**Comment by zimbatm** [comment]
is looks like it's the same issue when running `rm .envrc && ln -s .envrc--staging .envrc` on the same line.

direnv's main mechanism to detect file changes is by looking at the mtime, but that seems to be changing.

**Comment by zimbatm** [comment]
you can see in the direnv status that the found and loaded .envrc have the same timestamp. I don't know exactly why, probably due to the symlink, but that's what's preventing the reload

**Comment by brikou** [comment]
Something prevent direnv to be aware that .envrc has been removed, like if the pointer to this file is still the same.

**Comment by brikou** [comment]
Can you reproduce this issue on your system?

**Comment by brikou** [comment]
@zimbatm you are right, I confirm `rm .envrc && ln -s .envrc--staging .envrc` does not work, whereas `rm .envrc; ln -s .envrc--staging .envrc` do

**Comment by zimbatm** [comment]
@brikou I have implemented a fix in master, let me know how it goes!

**Comment by brikou** [comment]
Here is the script I ran to check direnv against master...

First download direnv and build it

```sh
cd $(mktemp -d)

docker container run \
    --rm \
    --volume=$PWD:/go/bin \
    golang \
        go get github.com/direnv/direnv

sudo mv direnv /usr/local/bin
```

Then run my use case...

```
cd $(mktemp -d)

cat << 'EOF' > .envrc--production
export ENV=production
EOF

cat << 'EOF' > .envrc--development
export ENV=development
EOF

ln --symbolic --force .envrc--development .envrc

direnv allow

echo $ENV
# development

ln --symbolic --force .envrc--production .envrc
# no red warning to allow this new conf

echo $ENV
# still development :/
```

Sadly it is still no working :/

Can you double check this script (by reading and then running it) and tell me what you get?

Thank you for your time on this issue!

**Comment by zimbatm** [comment]
can you `touch .envrc--production` so they have different timestamps and try switching again?

**Comment by brikou** [comment]
If I touch .env--production to make it older, it does not work neither. :/

**Comment by brikou** [comment]
@zimbatm can you reopen this issue?

**Comment by zimbatm** [comment]
I can't reproduce the issue on my machine with the latest release. But it's on macOS, maybe the filesystem is different.

**Comment by brikou** [comment]
Maybe you can try with virtual box (or vagrant) ?

**Comment by wbolster** [comment]
perhaps direnv should use `LStat` instead of `Stat`? (or both)

https://golang.org/src/os/stat.go

**Comment by zimbatm** [comment]
this is fixed by #452 :tada:
